File: tree/traversal/level_traversal/level_traversal.go
package level_traversal

import (
	"trouble_solver/tree"
)

// 层序遍历
func levelTraversal(root *tree.TreeNode) [][]int {
	if root == nil {
		return [][]int{}
	}

	var ans [][]int
	q := tree.Queue[tree.TreeNode]{}
	q.Push(root)

	tmp := []int{}
	for !q.Empty() {
		size := q.Length()
		for size > 0 {
			node := q.Pop()
			tmp = append(tmp, node.Val)
			size--

			if node.Left != nil {
				q.Push(node.Left)
			}
			if node.Right != nil {
				q.Push(node.Right)
			}
		}
		ans = append(ans, tmp)
		// 清空tmp
		tmp = []int{}
	}
	return ans
}

func levelTraversalRecursive(root *tree.TreeNode) [][]int {
	ans := [][]int{}
	depth := 0

	var order func(root *tree.TreeNode, depth int)
	order = func(root *tree.TreeNode, depth int) {
		if root == nil {
			return
		}
		if len(ans) == depth {
			ans = append(ans, []int{})
		}
		ans[depth] = append(ans[depth], root.Val)

		order(root.Left, depth+1)
		order(root.Right, depth+1)
	}

	order(root, depth)

	return ans
}

File: tree/traversal/level_traversal/LevelTraversal.java
/*
 * @Author: yangxcc
 * @version: 1.0
 * @Date: 2022-10-23 16:26:45
 * @LastEditTime: 2023-01-19 14:16:39
 */
package tree.traversal.level_traversal;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;

import tree.TreeNode;

/**
 * leetcode 102 middle 二叉树的层序遍历
 */
public class LevelTraversal {
    public List<List<Integer>> levelTraverse(TreeNode root) {
        List<List<Integer>> ans = new ArrayList<>();
        if (root == null) {
            return ans;
        }
        Deque<TreeNode> queue = new ArrayDeque<>();
        queue.offer(root);

        while (!queue.isEmpty()) {
            int size = queue.size();
            List<Integer> level = new ArrayList<>();
            while (size > 0) {
                TreeNode node = queue.poll();
                if (node.left != null) {
                    queue.offer(node.left);
                }
                if (node.right != null) {
                    queue.offer(node.right);
                }
                level.add(node.val);
                size--;
            }
            ans.add(new ArrayList<>(level));
        }

        return ans;
    }
}